### ISSUE TITLE
'Zoom out' is inconsistent on Mac

### ISSUE THREAD

**Issue by bskendig** [mrkite/minutor#319]
I downloaded Minutor 2.19.0 (Minutor.macOS-5.12.zip). (Get Info on the app icon still shows it as version 2.18.0, btw.) I deleted `~/Library/Application Support/seancode`. Launched the app and opened my 1.19.0 world.

With my world displayed in the default-sized window:
* The mouse scroll wheel zooms the map in/out normally.
* If I'm zoomed all the way out, holding the Command key and scrolling down will zoom the map out one more level.
* Releasing the Command key and trying to zoom out further will actually zoom in one level, setting the zoom level back to the maximum that was allowed without the Command key.

If I drag the corner of the window to make it larger, at some point (roughly double the area of the original window) it will stop allowing me to use the Command key to zoom out that extra level. This kind of defeats the purpose of being able to zoom out further; it doesn't let me see more of the map if I'm limited to a small map window.

If I had the smaller map zoomed out that extra level, then the app 'remembers' this. As I resize the map window, it will automatically zoom out one level when the window is small enough, and back in one level when I grow the window larger. I can keep the extra zoom-out level with a very tall-but-narrow or fat-but-short window, which makes me suspect it's limited by the total area displayed.

I just noticed that shrinking the window to the smallest size that I can shrink it to will let me zoom out _two_ levels. But zooming out two levels in a tiny window really isn't useful.

In the Minutor preferences, setting "Zoom Out" to "Ctrl" lets the Command (not Ctrl) key control that extra zoom level. Setting it to "Alt" lets the Option key control it. Setting it to "Shift" does not let the Shift key control it, though.

Similarly, setting "Depth Slider" to "Ctrl" lets the Command key control the depth setting, and setting it to "Alt" lets the Option key control depth. Setting it to "Shift" does not let the Shift key control it (no effect when Shift-scrolling). I noticed that while Command- or Option-scrolling to set the depth, the number above the right-hand slider changes but the slider position itself does not update to match the number.

Ideally, I'd love to have a large (fullscreen) window and be able to zoom out several more levels. I know this is memory-intensive, but Macs these days can have a lot of memory. You could also implement this the way that the "uNmINeD" Minecraft mapper does it, by not showing terrain detail above a certain zoom-out level, only grey squares to represent what chunks exist, so that a user can find distant chunks and zoom in on them.

**Comment by EtlamGit** [comment]
Regarding the keys/key-bindings: Can you post an image of an typical (up to date) Mac keyboard and explain how these extra keys are used in other Apps? (I have no clue about MacOS)

The zoom behavior is exactly what is is intended to be. There is massive amount of memory needed for all that Chunks, as we are rendering much more than Minecraft will do ingame. The allowed zoom level is restricted by your available RAM and actual window size. Therefore you can zoom out in small window, but not in large one because that will need too much Chunks for your Computer. In extreme cases you will even not be allowed to zoom to 1:1 in larger windows.

I see some possible reasons:
* Not enough RAM available: (Buy more) Close all other apps before starting Minutor. The RAM has to be physically available, no swap.
* RAM is not detected: It might be that we do not detect your RAM size correctly. We would have to debug that case.

**Comment by bskendig** [comment]
Here's an image of the keyboard of an M1 MacBook Pro (16-inch), the latest model:

https://www.apple.com/v/macbook-pro-14-and-16/b/images/overview/keyboard/keyboard__0zc8gxnp2gy2_large.jpg

The Command key is used for similar functions to what Windows uses the Ctrl key for. ⌘Q is Quit, ⌘W is Close, ⌘Z is Undo, ⌘X is Cut, and so forth.

Minutor's use of the Command and Options keys (instead of Ctrl and Alt) is correct. All you should do is update the text of the preferences to use the correct key names on Mac.

**Comment by bskendig** [comment]
I understand what you're saying about RAM limitations for Minutor. It would be nice to have an option to try to continue to use more RAM without limitation, though, for people who like to live on the edge. Or to have Minutor at a zoomed-out level simply show me allocated chunks (not any details of what's on them) so that I could get a 'big picture' of my world and then zoom in on part of it to see the detail (that's what uNmiNeD does).

Actually, it looks like Minutor might be reading my available RAM incorrectly. I'm running it on a Mac mini with (only) 8 GB RAM. When I load my world, make the window fullscreen (2560x1440), and zoom out as far as I can, the most memory I've been able to get it to use (unless I'm reading Activity Monitor wrong) has been 1.44 GB.

**Comment by EtlamGit** [comment]
I have a similar problem when trying to use ALT as modifier for depth slider.
On Windows ALT will change the vertical scroll-wheel into a horizontal one. With that the vertical change value will be zero and therefore it looks like it would not work. This should be solvable by combining `angleDelta().y()` and` angleDelta().x()`.

**Comment by EtlamGit** [comment]
regarding RAM detection: are you able to compile Minutor yourself?

**Comment by bskendig** [comment]
With the instructions on https://github.com/mrkite/minutor/wiki/How-to-Build-yourself-from-source, I can try. Are those instructions up-to-date; does it still use Qt 5.12?

**Comment by EtlamGit** [comment]
You should be able to use any Qt 5.x version. Try with the one shipped with your macOS, that might be easier.

Do NOT start with a static build like in the instructions you mentioned. That way is too complicated, as you have to build Qt itself. This is only necessary when you target for a single executable. It should be possible to install system Qt and then just call qmake/make. [Our automatic build flow for macOS](https://github.com/mrkite/minutor/blob/master/.github/workflows/cpp-qt-macos.yml) might give some hints.

**Comment by bskendig** [comment]
fyi, the macOS tool Homebrew (https://brew.sh) is generally used for installing open-source stuff.
`brew install qt5` gives me Qt 5.15.5.
Then `/usr/local/opt/qt@5/bin/qmake minutor.pro` succeeded.

There is no `debug` target in the minutor Makefile, so I just ran `make` and that built `minutor.app`.

I see the code you have for determining available memory is:
```
#if defined(__unix__) || defined(__unix) || defined(unix)
#ifdef _SC_AVPHYS_PAGES
  auto pages = sysconf(_SC_AVPHYS_PAGES);
  auto page_size = sysconf(_SC_PAGE_SIZE);
  quint64 available = (pages*page_size);
  chunks   = available / sizeChunkMax;
  maxcache = available / sizeChunkTypical;  // most chunks are less filled with sections                                
#endif
```

But I see a different approach on https://stackoverflow.com/questions/30511579/linux-sc-phys-pages-not-working-on-mac-os-x. I'll look into it.

**Comment by bskendig** [comment]
Question - what line of code can I add to display the value of `available` somewhere where I can see it? Like, maybe in the status bar?

**Comment by bskendig** [comment]
I believe I've got a fix that properly reads the amount of physical memory on macOS. (The only caveat is that I can only read the total physical memory, not free memory, so using the total physical memory will mean that swap will be used because Minutor isn't the only thing occupying physical memory.)

If you'll give me permission to push a branch up, I'll open a pull request for you.

**Comment by EtlamGit** [comment]
In `mapview.cpp` at lines 606 - 611 you will find the content of the status bar that is displayed only in DEBUG mode.
Just add a true to the condition.

With that you can see how big the ChunkCache is at maximum (and fill level).

**Comment by EtlamGit** [comment]
Regarding physical vs available memory:
* It is absolutely important that we do not use more memory than available. Minutor will freeze (and may crash) in this case.
* As not all Chunks are filled with Blocks to build limit, so we have some margin. But this is not deterministic.
* There are other sources for memory usage: Entities, Structures
* Other apps are not that big of a problem. They just get swapped out and as long as you are not using them at the same time it works.

To conclude: We may have to twiddle the ChuckCache size a bit to work smooth on macOS.

**Comment by EtlamGit** [comment]
https://superuser.com/questions/197059/mac-os-x-sysctl-get-total-and-free-memory-size
https://nxmnpg.lemoda.net/3/sysctlbyname
https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/sysctlbyname.3.html

**Comment by bskendig** [comment]
The challenge with getting the amount of free memory is that it could change greatly and frequently, though, right? If you launch Minutor on an 8GB system there might only be 2GB free, but as Minutor claims more memory, other apps will get pushed out to swap space and Minutor will have more real memory to work with. But you'll never be allowed to use the entire 8GB, as some amount of that will always be used by the operating system and other software that needs to remain resident.

In my testing with Minutor using 8GB on my Mac, the app became slow, but it didn't freeze or crash.

Here's a `git diff` of my changes if you'd like to add them or improve on them. At the very least this adds macOS-specific memory handling, because previously the check for `unix` wasn't catching it and it was falling through to the default memory management of using exactly 10000 chunks.

```
% git diff master
diff --git a/chunkcache.cpp b/chunkcache.cpp
index a120b94..da62e10 100644
--- a/chunkcache.cpp
+++ b/chunkcache.cpp
@@ -8,6 +8,9 @@
 #include <unistd.h>
 #elif defined(_WIN32) || defined(WIN32)
 #include <windows.h>
+#elif __APPLE__
+#include <unistd.h>
+#include <sys/sysctl.h>
 #endif
 
 ChunkCache::ChunkCache() {
@@ -34,6 +37,12 @@ ChunkCache::ChunkCache() {
   DWORDLONG available = qMin(status.ullAvailPhys, status.ullAvailVirtual);
   chunks   = available / sizeChunkMax;
   maxcache = available / sizeChunkTypical;  // most chunks are less filled with sections
+#elif __APPLE__
+  uint64_t available;
+  size_t len = sizeof(available);
+  sysctlbyname("hw.memsize", &available, &len, NULL, 0);
+  chunks   = available / sizeChunkMax;
+  maxcache = available / sizeChunkTypical;  // most chunks are less filled with sections
 #endif
   // we start the Cache based on worst case calculation
   cache.setMaxCost(chunks);
```

**Comment by EtlamGit** [comment]
Yes, current solution is somewhat gamble that the "free" memory stays similar to the moment Minutor was started.
Reasoning is, that you might have Minecraft open at the same time and switching between the apps. In that case we are not allowed to claim more memory.
Maybe we should place some percentage scaling factor, e.g. 80% of physical RAM.

Just create a PullRequest with your changes. We can start with your suggestion and if others report stability issues we reduce that factor.

**Comment by bskendig** [comment]
Looks great! I am now able to use more memory to zoom further out when holding down a modifier key, the Shift key works as a modifier (when it's selected, of course) for zooming and for the depth slider, and the depth slider handle moves up and down as I use the mouse wheel + modifier to scroll the depth. Thank you!

**Comment by EtlamGit** [comment]
In that case, I think we could close this issue now.

The only things we have to observe in future are:
* Does Minutor crash now on MacOS with out-of-memory. This will probably happen with multiple instances of Minutor. We might have to find a solution for that.
* How do dual-direction scroll wheels behave with my changes? I added X and Y coordinates, so vertical and horizontal scroll wheel are bound to each other.

**Comment by bskendig** [comment]
Sure, we can close this issue now. I'd really like a way to zoom out farther - either using arbitrarily more virtual memory at the cost of application speed, or by displaying gray blocks rather than rendering them above a certain zoom level (as uNmINeD does), but I understand that's not in Minutor's feature set.

Minutor does not run out of memory on macOS, because macOS will continue to increase swap space as needed. In theory if you're out of disk space it could cause problems, but that's not Minutor's problem to fix.

macOS doesn't support running multiple instances of an app. In theory you can if you launch it from Terminal by finding the executable inside the app bundle, but people aren't going to do that (and if they do then they deserve what they get).

On a Mac you can scroll horizontally by two-finger-dragging on a touchpad or by holding down Shift while scrolling. In theory the latter could interfere with Minutor's Shift preferences, but I just tested on my desktop (with only a traditional scroll-wheel mouse) and if Shift-scrolling was read as 'horizontal' then I didn't see any indication of it (or any idea of what horizontal scrolling was supposed to do).
